In [1]:
import pandas as pd
import numpy as np
import matplotlib as plt
import seaborn as sns

In [2]:
%matplotlib inline
pd.set_option('display.max_columns', None)

In [3]:
# import excel file Nov2020PrecinctDetail.xlsx
tn_2020=pd.read_excel('../data/2020_election/Nov2020PrecinctDetail.xlsx',
                      usecols=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE','RNAME1','PARTY1','PVTALLY1',
                             'RNAME2','PARTY2','PVTALLY2','RNAME3','PARTY3','PVTALLY3','RNAME4','PARTY4','PVTALLY4',
                             'RNAME5','PARTY5','PVTALLY5','RNAME6','PARTY6','PVTALLY6','RNAME7','PARTY7','PVTALLY7',
                             'RNAME8','PARTY8','PVTALLY8','RNAME9','PARTY9','PVTALLY9','RNAME10','PARTY10','PVTALLY10'])
tn_2020

COUNTY       PRECINCT  \
0      Anderson  Andersonville   
1      Anderson  Andersonville   
2      Anderson  Andersonville   
3      Anderson  Andersonville   
4      Anderson  Andersonville   
...         ...            ...   
13124    Wilson           25-2   
13125    Wilson           25-2   
13126    Wilson           25-2   
13127    Wilson           25-2   
13128    Wilson           25-2   

                                              OFFICENAME   ELECTDATE  \
0                                United States President  2020-11-03   
1                                United States President  2020-11-03   
2                                   United States Senate  2020-11-03   
3                                   United States Senate  2020-11-03   
4      United States House of Representatives District 3  2020-11-03   
...                                                  ...         ...   
13124                               United States Senate  2020-11-03   
13125                               United States Senate  2020-11-03   
13126  United States House of Representatives District 6  2020-11-03   
13127     Tennessee House of Representatives District 46  2020-11-03   
13128     Tennessee House of Representatives District 57  2020-11-03   

                         RNAME1       PARTY1  PVTALLY1  \
0               Donald J. Trump   Republican      1554   
1      Write-In - Brian Carroll          NaN         0   
2                  Bill Hagerty   Republican      1529   
3       Eric William Stansberry  Independent         4   
4             Chuck Fleischmann   Republican      1507   
...                         ...          ...       ...   
13124              Bill Hagerty   Republican      1224   
13125   Eric William Stansberry  Independent         6   
13126                 John Rose   Republican      1216   
13127                Clark Boyd   Republican       857   
13128              Susan M Lynn   Republican       486   

                          RNAME2       PARTY2  PVTALLY2  \
0                Joseph R. Biden   Democratic     307.0   
1        Write-In - Tom Hoefling          NaN       0.0   
2              Marquita Bradshaw   Democratic     268.0   
3      Write-In - John A. Gentry          NaN       0.0   
4                     Meg Gorman   Democratic     265.0   
...                          ...          ...       ...   
13124          Marquita Bradshaw   Democratic     440.0   
13125  Write-In - John A. Gentry          NaN       0.0   
13126  Christopher Martin Finley   Democratic     427.0   
13127                        NaN          NaN       NaN   
13128                 Tom Sottek  Independent     128.0   

                          RNAME3       PARTY3  PVTALLY3  \
0                Don Blankenship  Independent       1.0   
1        Write-In - Jade Simmons          NaN       0.0   
2      Yomi "Fapas" Faparusi Sr.  Independent       2.0   
3            Write-In - Al Green          NaN       0.0   
4                Amber L. Hysell  Independent      17.0   
...                          ...          ...       ...   
13124  Yomi "Fapas" Faparusi Sr.  Independent       5.0   
13125        Write-In - Al Green          NaN       0.0   
13126      Christopher B. Monday  Independent      47.0   
13127                        NaN          NaN       NaN   
13128                        NaN          NaN       NaN   

                           RNAME4       PARTY4  PVTALLY4  \
0      Roque "Rocky" De La Fuente  Independent       0.0   
1          Write-In - Kasey Wells          NaN       0.0   
2             Jeffrey Alan Grunau  Independent       3.0   
3                             NaN          NaN       NaN   
4          Keith Douglas Sweitzer  Independent      12.0   
...                           ...          ...       ...   
13124         Jeffrey Alan Grunau  Independent       2.0   
13125                         NaN          NaN       NaN   
13126                         NaN          NaN       NaN   
13127               

In [4]:
tn_2020.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13129 entries, 0 to 13128
Data columns (total 34 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   COUNTY      13129 non-null  object 
 1   PRECINCT    13129 non-null  object 
 2   OFFICENAME  13129 non-null  object 
 3   ELECTDATE   13129 non-null  object 
 4   RNAME1      13129 non-null  object 
 5   PARTY1      11164 non-null  object 
 6   PVTALLY1    13129 non-null  int64  
 7   RNAME2      11742 non-null  object 
 8   PARTY2      7504 non-null   object 
 9   PVTALLY2    11742 non-null  float64
 10  RNAME3      9738 non-null   object 
 11  PARTY3      5519 non-null   object 
 12  PVTALLY3    9738 non-null   float64
 13  RNAME4      7387 non-null   object 
 14  PARTY4      4854 non-null   object 
 15  PVTALLY4    7387 non-null   float64
 16  RNAME5      4511 non-null   object 
 17  PARTY5      3927 non-null   object 
 18  PVTALLY5    4511 non-null   float64
 19  RNAME6      3927 non-null

In [5]:
# merge entries in all RNAME columns into new column "name"
tn_2020_namemelt=tn_2020.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['RNAME1','RNAME2','RNAME3','RNAME4','RNAME5','RNAME6','RNAME7','RNAME8','RNAME9','RNAME10'],value_name="name")
# drop column "variable" created by .melt
tn_2020_namemelt=tn_2020_namemelt.drop(['variable'],axis=1)                          
tn_2020_namemelt.head(9405)

COUNTY             PRECINCT  \
0      Anderson        Andersonville   
1      Anderson        Andersonville   
2      Anderson        Andersonville   
3      Anderson        Andersonville   
4      Anderson        Andersonville   
...         ...                  ...   
9400  Robertson  01-1 East Robertson   
9401  Robertson  01-1 East Robertson   
9402  Robertson         02-1 Woodall   
9403  Robertson         02-1 Woodall   
9404  Robertson         02-1 Woodall   

                                             OFFICENAME   ELECTDATE  \
0                               United States President  2020-11-03   
1                               United States President  2020-11-03   
2                                  United States Senate  2020-11-03   
3                                  United States Senate  2020-11-03   
4     United States House of Representatives District 3  2020-11-03   
...                                                 ...         ...   
9400  United States House of Representatives District 6  2020-11-03   
9401     Tennessee House of Representatives District 66  2020-11-03   
9402                            United States President  2020-11-03   
9403                            United States President  2020-11-03   
9404                               United States Senate  2020-11-03   

                          name  
0              Donald J. Trump  
1     Write-In - Brian Carroll  
2                 Bill Hagerty  
3      Eric William Stansberry  
4            Chuck Fleischmann  
...                        ...  
9400                 John Rose  
9401          Sabi (Doc) Kumar  
9402           Donald J. Trump  
9403  Write-In - Brian Carroll  
9404              Bill Hagerty  

[9405 rows x 5 columns]

In [6]:
# merge entries in all PARTY columns into new column "party"
tn_2020_partymelt=tn_2020.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['PARTY1','PARTY2','PARTY3','PARTY4','PARTY5','PARTY6','PARTY7','PARTY8','PARTY9','PARTY10'],value_name="party")
# drop column "variable" created by .melt
tn_2020_partymelt=tn_2020_partymelt.drop(['variable'],axis=1)
tn_2020_partymelt.head(9405)

COUNTY             PRECINCT  \
0      Anderson        Andersonville   
1      Anderson        Andersonville   
2      Anderson        Andersonville   
3      Anderson        Andersonville   
4      Anderson        Andersonville   
...         ...                  ...   
9400  Robertson  01-1 East Robertson   
9401  Robertson  01-1 East Robertson   
9402  Robertson         02-1 Woodall   
9403  Robertson         02-1 Woodall   
9404  Robertson         02-1 Woodall   

                                             OFFICENAME   ELECTDATE  \
0                               United States President  2020-11-03   
1                               United States President  2020-11-03   
2                                  United States Senate  2020-11-03   
3                                  United States Senate  2020-11-03   
4     United States House of Representatives District 3  2020-11-03   
...                                                 ...         ...   
9400  United States House of Representatives District 6  2020-11-03   
9401     Tennessee House of Representatives District 66  2020-11-03   
9402                            United States President  2020-11-03   
9403                            United States President  2020-11-03   
9404                               United States Senate  2020-11-03   

            party  
0      Republican  
1             NaN  
2      Republican  
3     Independent  
4      Republican  
...           ...  
9400   Republican  
9401   Republican  
9402   Republican  
9403          NaN  
9404   Republican  

[9405 rows x 5 columns]

In [7]:
# merge entries in all PVTALLY columns into new column "vote_count"
tn_2020_pvtallymelt=tn_2020.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['PVTALLY1','PVTALLY2','PVTALLY3','PVTALLY4','PVTALLY5','PVTALLY6','PVTALLY7','PVTALLY8','PVTALLY9','PVTALLY10'],value_name="vote_count")
# drop column "variable" created by .melt
tn_2020_pvtallymelt=tn_2020_pvtallymelt.drop(['variable'],axis=1)
tn_2020_pvtallymelt.head(9405)

COUNTY             PRECINCT  \
0      Anderson        Andersonville   
1      Anderson        Andersonville   
2      Anderson        Andersonville   
3      Anderson        Andersonville   
4      Anderson        Andersonville   
...         ...                  ...   
9400  Robertson  01-1 East Robertson   
9401  Robertson  01-1 East Robertson   
9402  Robertson         02-1 Woodall   
9403  Robertson         02-1 Woodall   
9404  Robertson         02-1 Woodall   

                                             OFFICENAME   ELECTDATE  \
0                               United States President  2020-11-03   
1                               United States President  2020-11-03   
2                                  United States Senate  2020-11-03   
3                                  United States Senate  2020-11-03   
4     United States House of Representatives District 3  2020-11-03   
...                                                 ...         ...   
9400  United States House of Representatives District 6  2020-11-03   
9401     Tennessee House of Representatives District 66  2020-11-03   
9402                            United States President  2020-11-03   
9403                            United States President  2020-11-03   
9404                               United States Senate  2020-11-03   

      vote_count  
0         1554.0  
1            0.0  
2         1529.0  
3            4.0  
4         1507.0  
...          ...  
9400      2202.0  
9401      2153.0  
9402      2366.0  
9403         0.0  
9404      2328.0  

[9405 rows x 5 columns]

In [8]:
# add "party" and "vote_count" columns to tn_2020_namemelt
party=tn_2020_partymelt['party']
pvtally=tn_2020_pvtallymelt['vote_count']
tn_2020_namemelt=tn_2020_namemelt.join(party).join(pvtally)
tn_2020_namemelt.head(3)

COUNTY       PRECINCT               OFFICENAME   ELECTDATE  \
0  Anderson  Andersonville  United States President  2020-11-03   
1  Anderson  Andersonville  United States President  2020-11-03   
2  Anderson  Andersonville     United States Senate  2020-11-03   

                       name       party  vote_count  
0           Donald J. Trump  Republican      1554.0  
1  Write-In - Brian Carroll         NaN         0.0  
2              Bill Hagerty  Republican      1529.0

In [9]:
tn_2020_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 131290 entries, 0 to 131289
Data columns (total 7 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   COUNTY      131290 non-null  object 
 1   PRECINCT    131290 non-null  object 
 2   OFFICENAME  131290 non-null  object 
 3   ELECTDATE   131290 non-null  object 
 4   name        66142 non-null   object 
 5   party       50638 non-null   object 
 6   vote_count  66142 non-null   float64
dtypes: float64(1), object(6)
memory usage: 7.0+ MB


In [10]:
# lower case column names
tn_2020_namemelt.columns=tn_2020_namemelt.columns.str.lower()
tn_2020_namemelt.head(2)

county       precinct               officename   electdate  \
0  Anderson  Andersonville  United States President  2020-11-03   
1  Anderson  Andersonville  United States President  2020-11-03   

                       name       party  vote_count  
0           Donald J. Trump  Republican      1554.0  
1  Write-In - Brian Carroll         NaN         0.0

In [11]:
# change null values in "party" column to "no party"
tn_2020_namemelt['party'].fillna("No party", inplace = True)
tn_2020_namemelt.head(2)

county       precinct               officename   electdate  \
0  Anderson  Andersonville  United States President  2020-11-03   
1  Anderson  Andersonville  United States President  2020-11-03   

                       name       party  vote_count  
0           Donald J. Trump  Republican      1554.0  
1  Write-In - Brian Carroll    No party         0.0

In [12]:
tn_2020_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 131290 entries, 0 to 131289
Data columns (total 7 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   county      131290 non-null  object 
 1   precinct    131290 non-null  object 
 2   officename  131290 non-null  object 
 3   electdate   131290 non-null  object 
 4   name        66142 non-null   object 
 5   party       131290 non-null  object 
 6   vote_count  66142 non-null   float64
dtypes: float64(1), object(6)
memory usage: 7.0+ MB


In [13]:
# change null values in "vote_count" column to "0"
tn_2020_namemelt['vote_count'].fillna(0)

0         1554.0
1            0.0
2         1529.0
3            4.0
4         1507.0
           ...  
131285      10.0
131286       0.0
131287       0.0
131288       0.0
131289       0.0
Name: vote_count, Length: 131290, dtype: float64

In [14]:
tn_2020_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 131290 entries, 0 to 131289
Data columns (total 7 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   county      131290 non-null  object 
 1   precinct    131290 non-null  object 
 2   officename  131290 non-null  object 
 3   electdate   131290 non-null  object 
 4   name        66142 non-null   object 
 5   party       131290 non-null  object 
 6   vote_count  66142 non-null   float64
dtypes: float64(1), object(6)
memory usage: 7.0+ MB


# 66142 non-null names and 66142 non-null vote counts
# Do the nulls in "name" correlate one to one with "no party"?

In [15]:
tn_2020_namemelt[tn_2020_namemelt['name'].isna()]

county      precinct  \
13140   Anderson    Briceville   
13146   Anderson      Bull Run   
13152   Anderson       Clinton   
13158   Anderson  Clinton High   
13164   Anderson       Claxton   
...          ...           ...   
131284    Wilson          25-2   
131286    Wilson          25-2   
131287    Wilson          25-2   
131288    Wilson          25-2   
131289    Wilson          25-2   

                                               officename   electdate name  \
13140      Tennessee House of Representatives District 33  2020-11-03  NaN   
13146      Tennessee House of Representatives District 33  2020-11-03  NaN   
13152      Tennessee House of Representatives District 33  2020-11-03  NaN   
13158      Tennessee House of Representatives District 33  2020-11-03  NaN   
13164      Tennessee House of Representatives District 33  2020-11-03  NaN   
...                                                   ...         ...  ...   
131284                            United States President  2020-11-03  NaN   
131286                               United States Senate  2020-11-03  NaN   
131287  United States House of Representatives District 6  2020-11-03  NaN   
131288     Tennessee House of Representatives District 46  2020-11-03  NaN   
131289     Tennessee House of Representatives District 57  2020-11-03  NaN   

           party  vote_count  
13140   No party         NaN  
13146   No party         NaN  
13152   No party         NaN  
13158   No party         NaN  
13164   No party         NaN  
...          ...         ...  
131284  No party         NaN  
131286  No party         NaN  
131287  No party         NaN  
131288  No party         NaN  
131289  No party         NaN  

[65148 rows x 7 columns]

In [16]:
tn_2020_namemelt[tn_2020_namemelt['name'].isna()].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 65148 entries, 13140 to 131289
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   county      65148 non-null  object 
 1   precinct    65148 non-null  object 
 2   officename  65148 non-null  object 
 3   electdate   65148 non-null  object 
 4   name        0 non-null      object 
 5   party       65148 non-null  object 
 6   vote_count  0 non-null      float64
dtypes: float64(1), object(6)
memory usage: 4.0+ MB


### Looks safe to drop rows with no votes tallied

In [17]:
# drop rows with no name and vote count
tn_2020_final=tn_2020_namemelt.dropna()

In [18]:
tn_2020_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 66142 entries, 0 to 131285
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   county      66142 non-null  object 
 1   precinct    66142 non-null  object 
 2   officename  66142 non-null  object 
 3   electdate   66142 non-null  object 
 4   name        66142 non-null  object 
 5   party       66142 non-null  object 
 6   vote_count  66142 non-null  float64
dtypes: float64(1), object(6)
memory usage: 4.0+ MB


In [19]:
tn_2020_final

county       precinct  \
0       Anderson  Andersonville   
1       Anderson  Andersonville   
2       Anderson  Andersonville   
3       Anderson  Andersonville   
4       Anderson  Andersonville   
...          ...            ...   
131273    Wilson           24-1   
131277    Wilson           25-1   
131279    Wilson           25-1   
131283    Wilson           25-2   
131285    Wilson           25-2   

                                               officename   electdate  \
0                                 United States President  2020-11-03   
1                                 United States President  2020-11-03   
2                                    United States Senate  2020-11-03   
3                                    United States Senate  2020-11-03   
4       United States House of Representatives District 3  2020-11-03   
...                                                   ...         ...   
131273                               United States Senate  2020-11-03   
131277                            United States President  2020-11-03   
131279                               United States Senate  2020-11-03   
131283                            United States President  2020-11-03   
131285                               United States Senate  2020-11-03   

                            name        party  vote_count  
0                Donald J. Trump   Republican      1554.0  
1       Write-In - Brian Carroll     No party         0.0  
2                   Bill Hagerty   Republican      1529.0  
3        Eric William Stansberry  Independent         4.0  
4              Chuck Fleischmann   Republican      1507.0  
...                          ...          ...         ...  
131273              Kacey Morgan  Independent        14.0  
131277     Write-In - R19 Boddie     No party         0.0  
131279              Kacey Morgan  Independent        14.0  
131283     Write-In - R19 Boddie     No party         0.0  
131285              Kacey Morgan  Independent        10.0  

[66142 rows x 7 columns]

In [20]:
# convert vote_count data from float64 to int 32 to to drop ".0"
tn_2020_final=tn_2020_final.astype({'vote_count':'int32'})
tn_2020_final

county       precinct  \
0       Anderson  Andersonville   
1       Anderson  Andersonville   
2       Anderson  Andersonville   
3       Anderson  Andersonville   
4       Anderson  Andersonville   
...          ...            ...   
131273    Wilson           24-1   
131277    Wilson           25-1   
131279    Wilson           25-1   
131283    Wilson           25-2   
131285    Wilson           25-2   

                                               officename   electdate  \
0                                 United States President  2020-11-03   
1                                 United States President  2020-11-03   
2                                    United States Senate  2020-11-03   
3                                    United States Senate  2020-11-03   
4       United States House of Representatives District 3  2020-11-03   
...                                                   ...         ...   
131273                               United States Senate  2020-11-03   
131277                            United States President  2020-11-03   
131279                               United States Senate  2020-11-03   
131283                            United States President  2020-11-03   
131285                               United States Senate  2020-11-03   

                            name        party  vote_count  
0                Donald J. Trump   Republican        1554  
1       Write-In - Brian Carroll     No party           0  
2                   Bill Hagerty   Republican        1529  
3        Eric William Stansberry  Independent           4  
4              Chuck Fleischmann   Republican        1507  
...                          ...          ...         ...  
131273              Kacey Morgan  Independent          14  
131277     Write-In - R19 Boddie     No party           0  
131279              Kacey Morgan  Independent          14  
131283     Write-In - R19 Boddie     No party           0  
131285              Kacey Morgan  Independent          10  

[66142 rows x 7 columns]

In [21]:
tn_2020_final.dtypes

county        object
precinct      object
officename    object
electdate     object
name          object
party         object
vote_count     int32
dtype: object

In [22]:
tn_2020_final['officename'].value_counts()

United States President                              27510
United States Senate                                 25506
United States House of Representatives District 3     1395
United States House of Representatives District 7     1100
United States House of Representatives District 8      976
                                                     ...  
Tennessee House of Representatives District 45          14
Tennessee House of Representatives District 99          12
Tennessee House of Representatives District 1           11
Tennessee House of Representatives District 68          11
Tennessee House of Representatives District 81           9
Name: officename, Length: 126, dtype: int64

In [23]:
tn_2020_final['officename'].unique()

array(['United States President', 'United States Senate',
       'United States House of Representatives District 3',
       'Tennessee House of Representatives District 36',
       'Tennessee House of Representatives District 33',
       'United States House of Representatives District 4',
       'Tennessee Senate District 14',
       'Tennessee House of Representatives District 62',
       'United States House of Representatives District 7',
       'Tennessee Senate District 24',
       'Tennessee House of Representatives District 75',
       'United States House of Representatives District 8',
       'Tennessee House of Representatives District 31',
       'United States House of Representatives District 2',
       'Tennessee Senate District 2',
       'Tennessee House of Representatives District 8',
       'Tennessee House of Representatives District 20',
       'Tennessee Senate District 10',
       'Tennessee House of Representatives District 24',
       'Tennessee House of Repre

In [24]:
tn_2020_final['vote_count'].value_counts()

0       22571
1        5980
2        4279
3        3195
4        2337
        ...  
2586        1
3382        1
4889        1
4430        1
2918        1
Name: vote_count, Length: 2726, dtype: int64

## 22571 rows have a vote count of 0. Drop those rows

In [25]:
tn_2020_final=tn_2020_final[tn_2020_final['vote_count'] != 0]
tn_2020_final=tn_2020_final.reset_index(drop=True)
tn_2020_final.head()

county       precinct                                         officename  \
0  Anderson  Andersonville                            United States President   
1  Anderson  Andersonville                               United States Senate   
2  Anderson  Andersonville                               United States Senate   
3  Anderson  Andersonville  United States House of Representatives District 3   
4  Anderson  Andersonville     Tennessee House of Representatives District 36   

    electdate                     name        party  vote_count  
0  2020-11-03          Donald J. Trump   Republican        1554  
1  2020-11-03             Bill Hagerty   Republican        1529  
2  2020-11-03  Eric William Stansberry  Independent           4  
3  2020-11-03        Chuck Fleischmann   Republican        1507  
4  2020-11-03            Dennis Powers   Republican        1528

In [26]:
# confirm no 0's in vote count
tn_2020_final['vote_count'].value_counts()

1       5980
2       4279
3       3195
4       2337
5       1869
        ... 
2322       1
2386       1
2983       1
2418       1
2049       1
Name: vote_count, Length: 2725, dtype: int64

In [27]:
# create data frame tn_2020_pres for US Senate election
tn_2020_uss_int=tn_2020_final[tn_2020_final['officename'].str.match('United States Senate')]
tn_2020_uss_int=tn_2020_uss_int.reset_index(drop=True)
tn_2020_uss_int

county       precinct            officename   electdate  \
0      Anderson  Andersonville  United States Senate  2020-11-03   
1      Anderson  Andersonville  United States Senate  2020-11-03   
2      Anderson     Briceville  United States Senate  2020-11-03   
3      Anderson     Briceville  United States Senate  2020-11-03   
4      Anderson       Bull Run  United States Senate  2020-11-03   
...         ...            ...                   ...         ...   
18628    Wilson           22-1  United States Senate  2020-11-03   
18629    Wilson           23-1  United States Senate  2020-11-03   
18630    Wilson           24-1  United States Senate  2020-11-03   
18631    Wilson           25-1  United States Senate  2020-11-03   
18632    Wilson           25-2  United States Senate  2020-11-03   

                          name        party  vote_count  
0                 Bill Hagerty   Republican        1529  
1      Eric William Stansberry  Independent           4  
2                 Bill Hagerty   Republican         317  
3      Eric William Stansberry  Independent           1  
4                 Bill Hagerty   Republican        1677  
...                        ...          ...         ...  
18628             Kacey Morgan  Independent          10  
18629             Kacey Morgan  Independent           7  
18630             Kacey Morgan  Independent          14  
18631             Kacey Morgan  Independent          14  
18632             Kacey Morgan  Independent          10  

[18633 rows x 7 columns]

In [28]:
tn_2020_uss_int.head(20)

county        precinct            officename   electdate  \
0   Anderson   Andersonville  United States Senate  2020-11-03   
1   Anderson   Andersonville  United States Senate  2020-11-03   
2   Anderson      Briceville  United States Senate  2020-11-03   
3   Anderson      Briceville  United States Senate  2020-11-03   
4   Anderson        Bull Run  United States Senate  2020-11-03   
5   Anderson        Bull Run  United States Senate  2020-11-03   
6   Anderson         Clinton  United States Senate  2020-11-03   
7   Anderson         Clinton  United States Senate  2020-11-03   
8   Anderson    Clinton High  United States Senate  2020-11-03   
9   Anderson    Clinton High  United States Senate  2020-11-03   
10  Anderson         Claxton  United States Senate  2020-11-03   
11  Anderson         Claxton  United States Senate  2020-11-03   
12  Anderson  Clinton Middle  United States Senate  2020-11-03   
13  Anderson  Clinton Middle  United States Senate  2020-11-03   
14  Anderson    Dutch Valley  United States Senate  2020-11-03   
15  Anderson    Emory Valley  United States Senate  2020-11-03   
16  Anderson    Emory Valley  United States Senate  2020-11-03   
17  Anderson        Fairview  United States Senate  2020-11-03   
18  Anderson        Fairview  United States Senate  2020-11-03   
19  Anderson     Glen Alpine  United States Senate  2020-11-03   

                       name        party  vote_count  
0              Bill Hagerty   Republican        1529  
1   Eric William Stansberry  Independent           4  
2              Bill Hagerty   Republican         317  
3   Eric William Stansberry  Independent           1  
4              Bill Hagerty   Republican        1677  
5   Eric William Stansberry  Independent           3  
6              Bill Hagerty   Republican        1513  
7   Eric William Stansberry  Independent           3  
8              Bill Hagerty   Republican         943  
9   Eric William Stansberry  Independent           3  
10             Bill Hagerty   Republican        1622  
11  Eric William Stansberry  Independent           3  
12             Bill Hagerty   Republican         530  
13  Eric William Stansberry  Independent           4  
14             Bill Hagerty   Republican         329  
15             Bill Hagerty   Republican        1145  
16  Eric William Stansberry  Independent           2  
17             Bill Hagerty   Republican         661  
18  Eric William Stansberry  Independent           3  
19             Bill Hagerty   Republican        1428

### Add column that has total vote count by count for use in calculating county vote percentages

In [29]:
# group county precincts into county per candidate, in descending order of votes within county
tn_2020_uss_int=tn_2020_uss_int.groupby(['electdate','officename','county','name','party'], as_index=False)['vote_count'].sum()
tn_2020_uss=tn_2020_uss_int.sort_values(by=['county','vote_count'],ascending=False).reset_index(drop=True)
tn_2020_uss.head(20)

electdate            officename      county                       name  \
0   2020-11-03  United States Senate      Wilson               Bill Hagerty   
1   2020-11-03  United States Senate      Wilson          Marquita Bradshaw   
2   2020-11-03  United States Senate      Wilson           Elizabeth McLeod   
3   2020-11-03  United States Senate      Wilson           Steven J. Hooper   
4   2020-11-03  United States Senate      Wilson                Aaron James   
5   2020-11-03  United States Senate      Wilson               Kacey Morgan   
6   2020-11-03  United States Senate      Wilson  Yomi "Fapas" Faparusi Sr.   
7   2020-11-03  United States Senate      Wilson              Ronnie Henley   
8   2020-11-03  United States Senate      Wilson    Eric William Stansberry   
9   2020-11-03  United States Senate      Wilson               G. Dean Hill   
10  2020-11-03  United States Senate      Wilson        Jeffrey Alan Grunau   
11  2020-11-03  United States Senate  Williamson               Bill Hagerty   
12  2020-11-03  United States Senate  Williamson          Marquita Bradshaw   
13  2020-11-03  United States Senate  Williamson           Elizabeth McLeod   
14  2020-11-03  United States Senate  Williamson  Yomi "Fapas" Faparusi Sr.   
15  2020-11-03  United States Senate  Williamson           Steven J. Hooper   
16  2020-11-03  United States Senate  Williamson              Ronnie Henley   
17  2020-11-03  United States Senate  Williamson                Aaron James   
18  2020-11-03  United States Senate  Williamson        Jeffrey Alan Grunau   
19  2020-11-03  United States Senate  Williamson               G. Dean Hill   

          party  vote_count  
0    Republican       50597  
1    Democratic       20059  
2   Independent         454  
3   Independent         323  
4   Independent         315  
5   Independent         308  
6   Independent         230  
7   Independent         211  
8   Independent         166  
9   Independent         109  
10  Independent          93  
11   Republican       91155  
12   Democratic       43432  
13  Independent        1315  
14  Independent         655  
15  Independent         295  
16  Independent         240  
17  Independent         168  
18  Independent         139  
19  Independent         133

In [30]:
# create new dataframe "county_vote" to use to find sum of votes per county
county_vote=tn_2020_uss.loc[0:,['county','vote_count']]
county_vote

county  vote_count
0       Wilson       50597
1       Wilson       20059
2       Wilson         454
3       Wilson         323
4       Wilson         315
...        ...         ...
1071  Anderson          58
1072  Anderson          55
1073  Anderson          54
1074  Anderson          53
1075  Anderson           1

[1076 rows x 2 columns]

In [31]:
#sum county votes
county_vote_sum=county_vote.groupby(['county'])['vote_count'].sum()
county_vote_sum

county
Anderson       34548
Bedford        18153
Benton          6983
Bledsoe         5473
Blount         64679
               ...  
Wayne           6168
Weakley        13348
White          11704
Williamson    137667
Wilson         72865
Name: vote_count, Length: 95, dtype: int32

In [32]:
# use "county_vote_sum" to add column "sum_county_vote" to tn_2020_pres correlating with value in each row
tn_2020_uss['sum_county_vote']=tn_2020_uss['county'].map(county_vote_sum)
tn_2020_uss.head(20)

electdate            officename      county                       name  \
0   2020-11-03  United States Senate      Wilson               Bill Hagerty   
1   2020-11-03  United States Senate      Wilson          Marquita Bradshaw   
2   2020-11-03  United States Senate      Wilson           Elizabeth McLeod   
3   2020-11-03  United States Senate      Wilson           Steven J. Hooper   
4   2020-11-03  United States Senate      Wilson                Aaron James   
5   2020-11-03  United States Senate      Wilson               Kacey Morgan   
6   2020-11-03  United States Senate      Wilson  Yomi "Fapas" Faparusi Sr.   
7   2020-11-03  United States Senate      Wilson              Ronnie Henley   
8   2020-11-03  United States Senate      Wilson    Eric William Stansberry   
9   2020-11-03  United States Senate      Wilson               G. Dean Hill   
10  2020-11-03  United States Senate      Wilson        Jeffrey Alan Grunau   
11  2020-11-03  United States Senate  Williamson               Bill Hagerty   
12  2020-11-03  United States Senate  Williamson          Marquita Bradshaw   
13  2020-11-03  United States Senate  Williamson           Elizabeth McLeod   
14  2020-11-03  United States Senate  Williamson  Yomi "Fapas" Faparusi Sr.   
15  2020-11-03  United States Senate  Williamson           Steven J. Hooper   
16  2020-11-03  United States Senate  Williamson              Ronnie Henley   
17  2020-11-03  United States Senate  Williamson                Aaron James   
18  2020-11-03  United States Senate  Williamson        Jeffrey Alan Grunau   
19  2020-11-03  United States Senate  Williamson               G. Dean Hill   

          party  vote_count  sum_county_vote  
0    Republican       50597            72865  
1    Democratic       20059            72865  
2   Independent         454            72865  
3   Independent         323            72865  
4   Independent         315            72865  
5   Independent         308            72865  
6   Independent         230            72865  
7   Independent         211            72865  
8   Independent         166            72865  
9   Independent         109            72865  
10  Independent          93            72865  
11   Republican       91155           137667  
12   Democratic       43432           137667  
13  Independent        1315           137667  
14  Independent         655           137667  
15  Independent         295           137667  
16  Independent         240           137667  
17  Independent         168           137667  
18  Independent         139           137667  
19  Independent         133           137667

In [33]:
# create "pct_county_vote" column for tn_2020_pres
tn_2020_uss['pct_county_vote']=round(tn_2020_uss['vote_count']/tn_2020_uss['sum_county_vote']*100,2)
tn_2020_uss

electdate            officename    county                       name  \
0     2020-11-03  United States Senate    Wilson               Bill Hagerty   
1     2020-11-03  United States Senate    Wilson          Marquita Bradshaw   
2     2020-11-03  United States Senate    Wilson           Elizabeth McLeod   
3     2020-11-03  United States Senate    Wilson           Steven J. Hooper   
4     2020-11-03  United States Senate    Wilson                Aaron James   
...          ...                   ...       ...                        ...   
1071  2020-11-03  United States Senate  Anderson        Jeffrey Alan Grunau   
1072  2020-11-03  United States Senate  Anderson               G. Dean Hill   
1073  2020-11-03  United States Senate  Anderson                Aaron James   
1074  2020-11-03  United States Senate  Anderson           Steven J. Hooper   
1075  2020-11-03  United States Senate  Anderson  Write-In - John A. Gentry   

            party  vote_count  sum_county_vote  pct_county_vote  
0      Republican       50597            72865            69.44  
1      Democratic       20059            72865            27.53  
2     Independent         454            72865             0.62  
3     Independent         323            72865             0.44  
4     Independent         315            72865             0.43  
...           ...         ...              ...              ...  
1071  Independent          58            34548             0.17  
1072  Independent          55            34548             0.16  
1073  Independent          54            34548             0.16  
1074  Independent          53            34548             0.15  
1075     No party           1            34548             0.00  

[1076 rows x 8 columns]

In [34]:
# how many parties?
tn_2020_uss['party'].value_counts()

Independent    855
Republican      95
Democratic      95
No party        31
Name: party, dtype: int64

### Create dataframe "tn_2020_county_uss_pct"  with percent values per party per county

In [44]:
# create new dataframe with reduced columns values only
tn_2020_pct_county=tn_2020_uss.loc[0:,['electdate','county','party','pct_county_vote']]
tn_2020_pct_county

electdate    county        party  pct_county_vote
0     2020-11-03    Wilson   Republican            69.44
1     2020-11-03    Wilson   Democratic            27.53
2     2020-11-03    Wilson  Independent             0.62
3     2020-11-03    Wilson  Independent             0.44
4     2020-11-03    Wilson  Independent             0.43
...          ...       ...          ...              ...
1071  2020-11-03  Anderson  Independent             0.17
1072  2020-11-03  Anderson  Independent             0.16
1073  2020-11-03  Anderson  Independent             0.16
1074  2020-11-03  Anderson  Independent             0.15
1075  2020-11-03  Anderson     No party             0.00

[1076 rows x 4 columns]

In [45]:
# create new dataframe with Republican values only
tn_2020_rep_pct=tn_2020_pct_county[tn_2020_pct_county['party']== "Republican"].reset_index(drop=True)
tn_2020_rep_pct

electdate      county       party  pct_county_vote
0   2020-11-03      Wilson  Republican            69.44
1   2020-11-03  Williamson  Republican            66.21
2   2020-11-03       White  Republican            80.96
3   2020-11-03     Weakley  Republican            75.90
4   2020-11-03       Wayne  Republican            86.75
..         ...         ...         ...              ...
90  2020-11-03      Blount  Republican            73.27
91  2020-11-03     Bledsoe  Republican            81.49
92  2020-11-03      Benton  Republican            78.22
93  2020-11-03     Bedford  Republican            76.68
94  2020-11-03    Anderson  Republican            66.54

[95 rows x 4 columns]

In [46]:
# rename 'pct_county_vote' to 'rep_pct' and drop "party" column to prepare for merge 
tn_2020_rep_pct=tn_2020_rep_pct.rename(columns={'pct_county_vote':'Republican'},inplace=False).drop(['party'], axis=1)
tn_2020_rep_pct

electdate      county  Republican
0   2020-11-03      Wilson       69.44
1   2020-11-03  Williamson       66.21
2   2020-11-03       White       80.96
3   2020-11-03     Weakley       75.90
4   2020-11-03       Wayne       86.75
..         ...         ...         ...
90  2020-11-03      Blount       73.27
91  2020-11-03     Bledsoe       81.49
92  2020-11-03      Benton       78.22
93  2020-11-03     Bedford       76.68
94  2020-11-03    Anderson       66.54

[95 rows x 3 columns]

In [47]:
# create new dataframe with Democratic values only
tn_2020_dem_pct=tn_2020_pct_county[tn_2020_pct_county['party']== "Democratic"].reset_index(drop=True)
tn_2020_dem_pct

electdate      county       party  pct_county_vote
0   2020-11-03      Wilson  Democratic            27.53
1   2020-11-03  Williamson  Democratic            31.55
2   2020-11-03       White  Democratic            17.02
3   2020-11-03     Weakley  Democratic            20.12
4   2020-11-03       Wayne  Democratic            10.91
..         ...         ...         ...              ...
90  2020-11-03      Blount  Democratic            24.66
91  2020-11-03     Bledsoe  Democratic            16.41
92  2020-11-03      Benton  Democratic            19.16
93  2020-11-03     Bedford  Democratic            20.98
94  2020-11-03    Anderson  Democratic            31.32

[95 rows x 4 columns]

In [48]:
# rename 'pct_county_vote' to 'dem_pct' and drop "party" to prepare for merge 
tn_2020_dem_pct=tn_2020_dem_pct.rename(columns={'pct_county_vote':'Democratic'},inplace=False).drop(['party'], axis=1)
tn_2020_dem_pct

electdate      county  Democratic
0   2020-11-03      Wilson       27.53
1   2020-11-03  Williamson       31.55
2   2020-11-03       White       17.02
3   2020-11-03     Weakley       20.12
4   2020-11-03       Wayne       10.91
..         ...         ...         ...
90  2020-11-03      Blount       24.66
91  2020-11-03     Bledsoe       16.41
92  2020-11-03      Benton       19.16
93  2020-11-03     Bedford       20.98
94  2020-11-03    Anderson       31.32

[95 rows x 3 columns]

In [49]:
# create new dataframe with non Republican or Democratic values only
tn_2020_other_pct=tn_2020_pct_county[(tn_2020_pct_county['party']=="Independent")|(tn_2020_pct_county['party']=="No party")].reset_index(drop=True)
tn_2020_other_pct

electdate    county        party  pct_county_vote
0    2020-11-03    Wilson  Independent             0.62
1    2020-11-03    Wilson  Independent             0.44
2    2020-11-03    Wilson  Independent             0.43
3    2020-11-03    Wilson  Independent             0.42
4    2020-11-03    Wilson  Independent             0.32
..          ...       ...          ...              ...
881  2020-11-03  Anderson  Independent             0.17
882  2020-11-03  Anderson  Independent             0.16
883  2020-11-03  Anderson  Independent             0.16
884  2020-11-03  Anderson  Independent             0.15
885  2020-11-03  Anderson     No party             0.00

[886 rows x 4 columns]

In [50]:
# sum by county
tn_2020_other_pct=tn_2020_other_pct.groupby(['county'],as_index=False)['pct_county_vote'].sum()
tn_2020_other_pct

county  pct_county_vote
0     Anderson             2.15
1      Bedford             2.35
2       Benton             2.61
3      Bledsoe             2.11
4       Blount             2.07
..         ...              ...
90       Wayne             2.33
91     Weakley             3.97
92       White             2.03
93  Williamson             2.24
94      Wilson             3.03

[95 rows x 2 columns]

In [51]:
# rename 'pct_county_vote' to 'other_pct' and drop "party" to prepare for merge 
tn_2020_other_pct=tn_2020_other_pct.rename(columns={'pct_county_vote':'Other'})
tn_2020_other_pct

county  Other
0     Anderson   2.15
1      Bedford   2.35
2       Benton   2.61
3      Bledsoe   2.11
4       Blount   2.07
..         ...    ...
90       Wayne   2.33
91     Weakley   3.97
92       White   2.03
93  Williamson   2.24
94      Wilson   3.03

[95 rows x 2 columns]

In [52]:
# join tn_2020_rep_pct, tn_2020_dem_pct and tn_2020_other_pct to form tn_2020_county_uss_pct
dem=tn_2020_dem_pct['Democratic']
other=tn_2020_other_pct['Other']
tn_2020_county_uss_pct=tn_2020_rep_pct.join(dem).join(other)
tn_2020_county_uss_pct

electdate      county  Republican  Democratic  Other
0   2020-11-03      Wilson       69.44       27.53   2.15
1   2020-11-03  Williamson       66.21       31.55   2.35
2   2020-11-03       White       80.96       17.02   2.61
3   2020-11-03     Weakley       75.90       20.12   2.11
4   2020-11-03       Wayne       86.75       10.91   2.07
..         ...         ...         ...         ...    ...
90  2020-11-03      Blount       73.27       24.66   2.33
91  2020-11-03     Bledsoe       81.49       16.41   3.97
92  2020-11-03      Benton       78.22       19.16   2.03
93  2020-11-03     Bedford       76.68       20.98   2.24
94  2020-11-03    Anderson       66.54       31.32   3.03

[95 rows x 5 columns]

### Add column for Republican margin over Democrat for heat mapping in Tableau

In [53]:
tn_2020_county_uss_pct['Rep_Margin']= tn_2020_county_uss_pct['Republican']-tn_2020_county_uss_pct['Democratic']
tn_2020_county_uss_pct.head(40)

electdate      county  Republican  Democratic  Other  Rep_Margin
0   2020-11-03      Wilson       69.44       27.53   2.15       41.91
1   2020-11-03  Williamson       66.21       31.55   2.35       34.66
2   2020-11-03       White       80.96       17.02   2.61       63.94
3   2020-11-03     Weakley       75.90       20.12   2.11       55.78
4   2020-11-03       Wayne       86.75       10.91   2.07       75.84
5   2020-11-03  Washington       68.20       29.37   2.63       38.83
6   2020-11-03      Warren       73.68       22.85   2.03       50.83
7   2020-11-03   Van Buren       79.93       17.36   2.47       62.57
8   2020-11-03       Union       83.67       14.29   2.24       69.38
9   2020-11-03      Unicoi       79.32       18.34   2.22       60.98
10  2020-11-03   Trousdale       75.28       22.25   2.41       53.03
11  2020-11-03      Tipton       75.07       22.21   2.18       52.86
12  2020-11-03      Sumner       70.70       26.78   1.79       43.92
13  2020-11-03    Sullivan       76.46       21.64   2.61       54.82
14  2020-11-03     Stewart       79.07       18.10   2.86       60.97
15  2020-11-03       Smith       78.77       17.79   3.59       60.98
16  2020-11-03      Shelby       36.12       61.05   1.89      -24.93
17  2020-11-03      Sevier       78.52       18.58   1.91       59.94
18  2020-11-03  Sequatchie       80.95       16.67   3.75       64.28
19  2020-11-03       Scott       87.49       10.09   2.94       77.40
20  2020-11-03  Rutherford       58.11       38.77   2.71       19.34
21  2020-11-03   Robertson       73.94       23.32   3.61       50.62
22  2020-11-03       Roane       75.21       21.56   2.59       53.65
23  2020-11-03        Rhea       81.62       16.14   2.66       65.48
24  2020-11-03      Putnam       71.77       25.76   1.87       46.01
25  2020-11-03        Polk       81.25       17.19   1.94       64.06
26  2020-11-03     Pickett       81.15       17.03   3.49       64.12
27  2020-11-03       Perry       81.68       16.14   2.60       65.54
28  2020-11-03     Overton       79.35       18.39   1.82       60.96
29  2020-11-03       Obion       79.58       17.47   2.66       62.11
30  2020-11-03      Morgan       84.51       13.67   2.49       70.84
31  2020-11-03       Moore       82.04       15.18   2.25       66.86
32  2020-11-03  Montgomery       56.54       39.21   1.92       17.33
33  2020-11-03      Monroe       80.86       17.05   1.56       63.81
34  2020-11-03       Meigs       80.62       16.94   3.65       63.68
35  2020-11-03     McNairy       80.77       16.12   2.66       64.65
36  2020-11-03      McMinn       80.28       17.65   2.28       62.63
37  2020-11-03       Maury       68.49       29.46   2.13       39.03
38  2020-11-03    Marshall       75.17       22.26   1.79       52.91
39  2020-11-03      Marion       74.54       23.15   3.44       51.39

In [54]:
#read TN_county_index_shape.csv to add index from TN Counties shape file for joining in Tableau
county_index=pd.read_csv('../data/TN_county_index_shape.csv')
county_index

county  INDEX_NO
0     Anderson        79
1      Bedford        26
2       Benton        41
3      Bledsoe        31
4       Blount        52
..         ...       ...
90       Wayne         6
91     Weakley        56
92       White        67
93  Williamson        44
94      Wilson        65

[95 rows x 2 columns]

In [55]:
# join with tn_2020_county_uss_pct
tn_2020_county_uss_pct=tn_2020_county_uss_pct.merge(county_index,on="county")
tn_2020_county_uss_pct

electdate      county  Republican  Democratic  Other  Rep_Margin  \
0   2020-11-03      Wilson       69.44       27.53   2.15       41.91   
1   2020-11-03  Williamson       66.21       31.55   2.35       34.66   
2   2020-11-03       White       80.96       17.02   2.61       63.94   
3   2020-11-03     Weakley       75.90       20.12   2.11       55.78   
4   2020-11-03       Wayne       86.75       10.91   2.07       75.84   
..         ...         ...         ...         ...    ...         ...   
89  2020-11-03      Blount       73.27       24.66   2.33       48.61   
90  2020-11-03     Bledsoe       81.49       16.41   3.97       65.08   
91  2020-11-03      Benton       78.22       19.16   2.03       59.06   
92  2020-11-03     Bedford       76.68       20.98   2.24       55.70   
93  2020-11-03    Anderson       66.54       31.32   3.03       35.22   

    INDEX_NO  
0         65  
1         44  
2         67  
3         56  
4          6  
..       ...  
89        52  
90        31  
91        41  
92        26  
93        79  

[94 rows x 7 columns]

In [56]:
tn_2020_county_uss_pct.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 94 entries, 0 to 93
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   electdate   94 non-null     object 
 1   county      94 non-null     object 
 2   Republican  94 non-null     float64
 3   Democratic  94 non-null     float64
 4   Other       94 non-null     float64
 5   Rep_Margin  94 non-null     float64
 6   INDEX_NO    94 non-null     int64  
dtypes: float64(4), int64(1), object(2)
memory usage: 5.9+ KB


In [57]:
# covert "electdate" to datetime from object
tn_2020_county_uss_pct['electdate']=pd.to_datetime(tn_2020_county_uss_pct['electdate'],format='%Y-%m-%d')
tn_2020_county_uss_pct

electdate      county  Republican  Democratic  Other  Rep_Margin  INDEX_NO
0  2020-11-03      Wilson       69.44       27.53   2.15       41.91        65
1  2020-11-03  Williamson       66.21       31.55   2.35       34.66        44
2  2020-11-03       White       80.96       17.02   2.61       63.94        67
3  2020-11-03     Weakley       75.90       20.12   2.11       55.78        56
4  2020-11-03       Wayne       86.75       10.91   2.07       75.84         6
..        ...         ...         ...         ...    ...         ...       ...
89 2020-11-03      Blount       73.27       24.66   2.33       48.61        52
90 2020-11-03     Bledsoe       81.49       16.41   3.97       65.08        31
91 2020-11-03      Benton       78.22       19.16   2.03       59.06        41
92 2020-11-03     Bedford       76.68       20.98   2.24       55.70        26
93 2020-11-03    Anderson       66.54       31.32   3.03       35.22        79

[94 rows x 7 columns]

In [58]:
tn_2020_county_uss_pct.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 94 entries, 0 to 93
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   electdate   94 non-null     datetime64[ns]
 1   county      94 non-null     object        
 2   Republican  94 non-null     float64       
 3   Democratic  94 non-null     float64       
 4   Other       94 non-null     float64       
 5   Rep_Margin  94 non-null     float64       
 6   INDEX_NO    94 non-null     int64         
dtypes: datetime64[ns](1), float64(4), int64(1), object(1)
memory usage: 5.9+ KB


In [59]:
tn_2020_county_uss_pct.to_excel(r'../data/2020_election/tn_2020_county_pres_pct.xlsx',index=False)